source/vistas/ui/windows/linreg_dialog.py

```diff
@@ -11,54 +11,374 @@
+            #Set bounds to fully zoomed out
+            self.ax.set_xlim(x_min, x_max)
+            self.ax.set_ylim(y_min, y_max)
+
+            #Save bound variables
+            self.x_lo = x_min
+            self.x_hi = x_max
+            self.y_lo = y_min
+            self.y_hi = y_max
+
+            #Calculate centerpoint
+            self.center_x = x_min + (x_max - x_min) / 2
+            self.center_y = y_min + (y_max - y_min) / 2
+
+            self.reset = False
+
+        # Current boundaries will be kept
+        keep_bounds = True
+
+        # Size of boundaries
+        x_length = (x_max - x_min)
+        y_length = (y_max - y_min)
+
+        # Divide sides for percentages
+        x_ticks = (x_length) / 100
+        y_ticks = (y_length) / 100
+
+        #USER DRAWN BOX TO ZOOM IN TO
+        if self.zoom_mode == 1:
+
+            # The length of the zoomed in bounds
+            x_length_current = abs(self.x_hi - self.x_lo)
+            y_length_current = abs(self.y_hi - self.y_lo)
+
+            # Percentage of the total bounds covered
+            x_percentage = self.mouse_x_diff / x_length
+            y_percentage = self.mouse_y_diff / y_length
+
+            # Figure out if the width or height is largest of the user's rectangle
+            if abs(x_percentage) >= abs(y_percentage):
+                zoom_value = ((x_length - abs(self.mouse_x_diff)) / 2) / x_ticks
+                x_box = (x_length_current * abs(x_percentage)) / 2
+                y_box = (y_length_current * abs(x_percentage)) / 2
+            else:
+                zoom_value = ((y_length - abs(self.mouse_y_diff)) / 2) / y_ticks
+                x_box = (x_length_current * abs(y_percentage)) / 2
+                y_box = (y_length_current * abs(y_percentage)) / 2
+
+            #DRAWING A BOX ON THE GRAPH
+            if self.draw:
+                square = Rectangle((0, 0), 1, 1, alpha=0.3, color='red')
+                self.ax.add_patch(square)
+
+                # Set length of box sides
+                square.set_width(x_box * 2)
+                square.set_height(y_box * 2)
+
+                # Draw corner of box depending on which direction the user is dragging
+                if (x_percentage >= 0) & (y_percentage >= 0):
+                    square.set_xy((self.mouse_x, self.mouse_y))
+                elif (x_percentage <= 0) & (y_percentage >= 0):
+                    square.set_xy((self.mouse_x - (x_box * 2), self.mouse_y))
+                elif (x_percentage <= 0) & (y_percentage <= 0):
+                    square.set_xy((self.mouse_x - (x_box * 2), self.mouse_y - (y_box * 2)))
+                else:
+                    square.set_xy((self.mouse_x, self.mouse_y - (y_box * 2)))
+
+                self.ax.figure.canvas.draw()
+
+            #ZOOMING INTO A DRAWN BOX
+            elif ((zoom_value < 50) & self.mouse_continue):
+
+                self.zoom.SetValue(round(zoom_value))
+
+                if (x_percentage >= 0) & (y_percentage >= 0):
```

Comment:
While `&` _technically_ results in the correct outcome here, it should be `and` instead. `&` is a bitwise operator, and aside from the clarity issue, with `&`, both sides are evaluated, whereas using `and` will stop evaluating if the first condition is `False`. For example:

```python
x = None

# Works because `x.foo` is never evaluated
if x is not None and x.foo == 2:
    ...

# Fails with an exception, because `x.foo` is evaluated when `x` is `None`:
if x is not None & x.foo == 2:
    ...
```

With that in mind, this condition should be:

```python
if x_percentage >=0 and y_percentage >= 0
    ...
```